src/monodroid/jni/marshal-methods-tracing.cc

```diff
@@ -0,0 +1,393 @@
+
+	size_t frame_counter = 0;
+
+	trace.append ("\n  Native stack trace:");
+	while (unw_step (&cursor) > 0) {
+		trace.append ("\n");
+
+		unw_get_reg (&cursor, UNW_REG_IP, &ip);
+		ip = adjust_address (ip);
+
+		auto ptr = reinterpret_cast<void*>(ip);
+		const char *fname = nullptr;
+		const void *symptr = nullptr;
+		unw_word_t  frame_offset = 0;
+		bool info_valid = false;
+
+		if (dladdr (ptr, &info) != 0) {
+			if (info.dli_fname != nullptr) {
+				fname = info.dli_fname;
+			}
+			symptr = info.dli_sname;
+			frame_offset = ip - reinterpret_cast<unw_word_t>(info.dli_fbase);
+			info_valid = true;
+		} else {
+			frame_offset = ip;
+		}
+
+		append_frame_number (trace, frame_counter++);
+
+		std::snprintf (num_buf.data (), num_buf.size (), "%0*zx (", FRAME_OFFSET_WIDTH, frame_offset);
+		trace.append (num_buf.data ());
+		std::snprintf (num_buf.data (), num_buf.size (), "%p) ", ptr);
+		trace.append (num_buf.data ());
+
+		// TODO: consider searching /proc/self/maps for the beginning of the corresponding region to calculate the
+		// correct offset (like done in bionic stack trace)
+		trace.append (fname != nullptr ? fname : "[anonymous]");
+
+		bool symbol_name_allocated = false;
+		offp = 0;
+		if (unw_get_proc_name (&cursor, name_buf.data (), name_buf.size (), &offp) == 0) {
+			symbol_name = name_buf.data ();
+		} else if (info_valid && info.dli_sname != nullptr) {
+			symbol_name = info.dli_sname;
+		} else {
+			symbol_name = nullptr;
+		}
+		offp = adjust_address (offp);
+
+		if (symbol_name != nullptr) {
+			char *demangled_symbol_name;
+			int demangle_status;
+
+			// https://itanium-cxx-abi.github.io/cxx-abi/abi.html#demangler
+			demangled_symbol_name = abi::__cxa_demangle (symbol_name, nullptr, nullptr, &demangle_status);
+			symbol_name_allocated = demangle_status == 0 && demangled_symbol_name != nullptr;
+			if (symbol_name_allocated) {
+				symbol_name = demangled_symbol_name;
+			}
+		}
+
+		if (symbol_name != nullptr) {
+			trace.append (" ");
+			trace.append (symbol_name);
+			if (offp != 0) {
+				trace.append (" + ");
+				std::snprintf (num_buf.data (), num_buf.size (), "%zu", offp);
+				trace.append (num_buf.data ());
+			}
+		}
+
+		if (symptr != nullptr) {
+			trace.append (" (symaddr: ");
+			std::snprintf (num_buf.data (), num_buf.size (), "%p", symptr);
+			trace.append (num_buf.data ());
+			trace.append (")");
+		}
+
+		if (symbol_name_allocated && symbol_name != nullptr) {
+			std::free (reinterpret_cast<void*>(const_cast<char*>(symbol_name)));
+		}
+	}
+}
+
+[[gnu::always_inline]]
+static void get_java_backtrace (JNIEnv *env, std::string &trace) noexcept
+{
+	// TODO: error handling
+	trace.append ("\n  Java stack trace:");
+	jobject current_thread = env->CallStaticObjectMethod (java_lang_Thread, java_lang_Thread_currentThread);
+	auto stack_trace_array = static_cast<jobjectArray>(env->CallNonvirtualObjectMethod (current_thread, java_lang_Thread, java_lang_Thread_getStackTrace));
```

Comment:
This should be `CallObjectMethod`, as [`java.lang.Thread.getStackTrace()`](https://developer.android.com/reference/java/lang/Thread#getStackTrace()) is a virtual (non-`final`) method.